pkg/fqdn/dnsproxy/udp.go

```diff
@@ -0,0 +1,250 @@
+	}
+}
+
+var udpOnce sync.Once
+
+func (s *sessionUDPFactory) InitConn(conn *net.UDPConn) error {
+	// Set up the raw socket for sending responses.
+	// - Must use a raw UDP socket for sending responses so that we can send
+	//   from a specific port without binding to it.
+	// - The raw UDP socket must be bound to a specific IP address to prevent
+	//   it receiving ALL UDP packets on the host.
+	// - We use oob data to override the source IP address when sending
+	// - Must use separate sockets for IPv4/IPv6, as sending to a v6-mapped
+	//   v4 address from a socket bound to "::1" does not work due to kernel
+	//   checking that a route exists from the source address before
+	//   the source address is replaced with the (transparently) changed one
+	udpOnce.Do(func() {
+		rawconn4 = bindUDP("127.0.0.1") // raw socket for sending IPv4
+		rawconn6 = bindUDP("::1")       // raw socket for sending IPv6
+	})
+	if rawconn4 == nil || rawconn6 == nil {
+		return fmt.Errorf("Unable to open raw UDP sockets for DNS Proxy")
+	}
+	return nil
+}
+
+// Clear re-initializes sessionUDP to the same state it was when new.
+// Returns the interface for convenience.
+// This is required for pooling of dns.SessionUDP objects.
+func (s *sessionUDP) Clear() dns.SessionUDP {
+	s.conn = nil
+	s.raddr = nil
+	s.laddr = nil
+	s.m = s.m[:cap(s.m)]
+	s.oob = s.oob[:cap(s.oob)]
+	return s
+}
+
+// RemoteAddr returns the remote network address.
+func (s *sessionUDP) RemoteAddr() net.Addr { return s.raddr }
+
+// LocalAddr returns the local network address for the current request.
+func (s *sessionUDP) LocalAddr() net.Addr { return s.laddr }
+
+// ReadRequest reads a single request from the session and keeps the request context
+func (s *sessionUDP) ReadRequest(conn *net.UDPConn) ([]byte, error) {
+	n, oobn, _, raddr, err := conn.ReadMsgUDP(s.m, s.oob)
+	if err == nil {
+		s.conn = conn
+		s.raddr = raddr
+		s.m = s.m[:n]        // Re-slice to the actual size
+		s.oob = s.oob[:oobn] // Re-slice to the actual size
+		s.laddr, err = parseDstFromOOB(s.oob)
+	}
+	return s.m, err
+}
+
+// WriteResponse writes a response to a request received earlier
+func (s *sessionUDP) WriteResponse(b []byte) (int, error) {
+	// Must give the UDP header to get the source port right.
+	// Reuse the msg buffer, figure out if golang can do gatter-scather IO
+	// with raw sockets?
+	bb := bytes.NewBuffer(s.m[:0])
+	binary.Write(bb, binary.BigEndian, uint16(s.laddr.Port))
+	binary.Write(bb, binary.BigEndian, uint16(s.raddr.Port))
+	binary.Write(bb, binary.BigEndian, uint16(8+len(b)))
+	binary.Write(bb, binary.BigEndian, uint16(0)) // checksum
+	bb.Write(b)
+	buf := bb.Bytes()
+
+	var n int
+	var err error
+	dst := net.IPAddr{
+		IP: s.raddr.IP,
+	}
+	if s.raddr.IP.To4() == nil {
+		n, _, err = rawconn6.WriteMsgIP(buf, s.controlMessage(s.laddr), &dst)
+	} else {
+		n, _, err = rawconn4.WriteMsgIP(buf, s.controlMessage(s.laddr), &dst)
+	}
+	if err != nil {
+		log.Fatalf("WriteMsgIP: %s", err)
```

Comment:
Why the fatal?

Reply:
Remnant from debugging, thanks for pointing this out :-)